src/circadian_model.R

```diff
@@ -51,41 +91,116 @@ circadianModel <- function(t, x, params){
     R1tot <- x[5]
 
     # Get R1b from R1tot and A
-    # TODO: R1b <- ...
+    R1b <- R1b <- 0.5 * (A + R1tot + 4 - sqrt(pow(A + R1tot + 4,2.0) - 4 * A * R1tot))
 
     # Determine if student is in school hours and if it is a school day
-    # TODO: inSchoolHours <- ...
-    # TODO: isSchoolDay <- ...
+    inSchoolHours <- (t %% 24 > schoolStartLocalTimeInHours) && (t %% 24 <= schoolStartLocalTimeInHours + schoolDurationInHours)
+    isSchoolDay <- ((t %/% 24) %% 7) < 5
 
     # Calculate the light intensity
-    # TODO: light <- ...
+    light <- lightApprox(t %% 24)$y
 
     # Determine if student is awake
-    # TODO: isAwake <- ...
+    if (inSchoolHours && isASchoolDay) {
+      isAwake <<- TRUE
+      light <- schoolBrightnessInLux
+    } else {
+      
+      # If it's free time, and you want to fall asleep, you can, as long as you're not up for social reasons
+      if (sleepDrive(R1b, Psi) > Dsleep) {  
+        if (t %% 24 > socialFactor || t %% 24 < 10) { # Forbid falling asleep before a certain time due to social reasons
+          isAwake <<- FALSE
+        }
+      }
+      # If it's free time, and you want to wake up, you can wake up
+      if (sleepDrive(R1b, Psi) < Dwake) {   
+        isAwake <<- TRUE
+      }
+    }
 
     # Set mu and ichi parameters based on whether student is awake or not
-    # TODO: mu <- ...
-    # TODO: ichi <- ...
+     
+    if (isAwake) {
+      ichi <- 1.0/18.18
+      mu <- 869.5
+      if (light < baselineLight) { # Set light to baselineLight if they're awake and it's dark outside
+        light <- baselineLight
+      }
+    } else {
+      ichi <- 1.0/7.0  # Originally 1.0/4.2
+      mu <- 596.5 
+      light <- 0  # Set light to 0 if they're asleep
+    }
 
     # Store lux value for plotting later
-    # TODO: light_schedule <- ...
+    if (allLux[1 + floor(t / dt)] == -1) {
+      allLux[1 + floor(t / dt)] <<- light
+    }
 
     # Calculate the right hand side of the ODEs
 
+    alpha_L <- alphaL(light)
+    BhatValue <- Bhat(N, light)
+    
     # Circadian model    
-    # dR = ...
-    # dPsi = ...
-    # dN = ...    
+    lightAmp <- BhatValue * amplitudeResponseCurve(R, Psi)
+    lightPhase <- BhatValue * phaseResponseCurve(R, Psi)
+    couplingTermAmp <- K * 0.5 * cos(Beta1) * R * (1.0 - pow(R,4))
+    
+    dR = (-1.0 * Gamma * R) + couplingTermAmp + lightAmp
+
+    naturalOscillation <- 2.0 * pi / Tau
+    couplingTermPhase <- K / 2.0 * sin(Beta1) * (1 + pow(R,4))
+    dPsi <- naturalOscillation + couplingTermPhase + lightPhase
+    
+    dN <- 60.0 * (alpha_L * (1 - N) - Delta * N)   
 
     # Sleep drive model
-    # dA = ...
-    # dR1tot = ...
+    dA <- ichi * (mu - A)
+    dR1tot <- Irecep * (R1b - Targc * R1tot)
 
     # Return the right hand side of the ODEs
-    # dxdt <- c(dR, dPsi, dN, dA, dR1tot)
+    dxdt <- c(dR, dPsi, dN, dA, dR1tot)
 
     # Return results as a list
-    # list(dxdt)
-    warning("circadianModel not implemented")
+    list(dxdt)
```

Comment:
This should be the end of the function (at `list(dxdt)`). All the code that's below should be something we define on the app and then call this function. The idea is for `circadianModel` to be a standalone function that we can call with different parameters later on. Basically the code you have below should be on the testing side because that's were you try your function out. Things to do:

- Move the code you have below to `test_model.R` and simulate it there.
- Make a couple of plots from the output to check that the simulation makes sense:
    - allLux vs. time
    - R vs. time
    - Psi vs. time
    -  N vs. time
    - A vs. time
    - R1tot vs. time
    -  Analyze if the time evolution of the quantities makes sense given the light input that the student had. For example, do they get the correct light when they are not in school? Are the circadian and sleep model interacting correctly?

Once we have tested that the function runs properly we can create a function called `simulateCircadian` were we give it all the parameters we want (like `numberOfDays`, `dt`, ...) and then get the output we want but for now let's do what I outlined above. Does that sound good?

Reply:
Got it, that makes sense. I assumed that the for loop for the sleepDurationSchool vector was to go in circadianModel, but I'll switch this over to the test_model.R.